File: apps/mobile/modules/nemu-aidoku/runtime/kotlin/NemuNativeHttpRequestHeaderPolicy.kt
package pm.nemu.mobile.aidoku

import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

/**
 * Bounds and canonicalizes source-authored request headers before OkHttp sees
 * them. Aidoku sources are untrusted packages, so malformed headers must fail
 * with a stable error instead of throwing from OkHttp's builder or allocating
 * an unbounded native header block.
 */
internal object NemuNativeHttpRequestHeaderPolicy {
  const val MAX_HEADER_COUNT = 128
  const val MAX_HEADER_NAME_CHARACTERS = 256
  const val MAX_HEADER_VALUE_CHARACTERS = 16 * 1024
  const val MAX_HEADER_VALUE_WIRE_BYTES = 16 * 1024
  const val MAX_TOTAL_HEADER_WIRE_BYTES = 64 * 1024

  private val tokenCharacters =
    "!#$%&'*+-.^_`|~0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".toSet()

  fun normalize(headers: Map<String, String>): Map<String, String> {
    if (headers.size > MAX_HEADER_COUNT) {
      throw IllegalArgumentException("Native HTTP request has too many headers.")
    }

    var totalWireBytes = 0
    val normalized = LinkedHashMap<String, String>(headers.size)
    val normalizedNames = HashSet<String>(headers.size)
    headers.forEach { (rawName, rawValue) ->
      // Retain the bridge's historical behavior for an empty property name.
      if (rawName.isBlank()) return@forEach
      if (
        rawName.length > MAX_HEADER_NAME_CHARACTERS ||
        rawName.any { it !in tokenCharacters }
      ) {
        throw IllegalArgumentException("Native HTTP request has an invalid header name.")
      }
      val normalizedName = rawName.lowercase()
      if (!normalizedNames.add(normalizedName)) {
        throw IllegalArgumentException("Native HTTP request has duplicate header names.")
      }
      if (rawValue.length > MAX_HEADER_VALUE_CHARACTERS) {
        throw IllegalArgumentException("Native HTTP request has an oversized header value.")
      }

      val value = normalizeUrlValue(normalizedName, rawValue)
      if (value.any { it != '\t' && (it < ' ' || it > '~') }) {
        throw IllegalArgumentException(
          "Native HTTP request header values must use printable ASCII."
        )
      }
      // Names and normalized values are ASCII at this point, so JVM string
      // length is the exact HTTP/1 field-content byte length. Rechecking after
      // URL canonicalization prevents percent-encoding from expanding a small
      // Unicode input past the advertised per-value bound.
      val valueWireBytes = value.length
      if (valueWireBytes > MAX_HEADER_VALUE_WIRE_BYTES) {
        throw IllegalArgumentException("Native HTTP request has an oversized header value.")
      }
      totalWireBytes += rawName.length + valueWireBytes
      if (totalWireBytes > MAX_TOTAL_HEADER_WIRE_BYTES) {
        throw IllegalArgumentException("Native HTTP request headers exceed the safety limit.")
      }
      // Sources never control Nemu's proxy credential. Validate and count the
      // field first so the reserved name cannot bypass bounds or case-folded
      // duplicate rejection.
      if (normalizedName == "proxy-authorization") return@forEach
      normalized[rawName] = value
    }
    return normalized
  }

  private fun normalizeUrlValue(normalizedName: String, value: String): String {
    if (value.all { it == '\t' || it in ' '..'~' }) return value
    if (
      normalizedName != "referer" &&
      normalizedName != "referrer" &&
      normalizedName != "origin"
    ) {
      return value
    }

    // HttpUrl serializes Unicode hosts and paths into an ASCII wire form.
    // This mirrors browser referrer behavior and avoids OkHttp rejecting the
    // request before it reaches the source site.
    return value.toHttpUrlOrNull()?.toString()
      ?: throw IllegalArgumentException("Native HTTP request has an invalid URL header.")
  }
}

File: apps/mobile/modules/nemu-aidoku/ios/NemuNativeHttpRequestHeaderPolicy.swift
import Foundation

/// Bounds and canonicalizes source-authored request headers before Foundation
/// sees them. Aidoku sources are untrusted packages, so malformed headers must
/// fail with a stable error instead of reaching URLSession or allocating an
/// unbounded native header block.
enum NemuNativeHttpRequestHeaderPolicy {
  static let maxHeaderCount = 128
  static let maxHeaderNameCharacters = 256
  static let maxHeaderValueCharacters = 16 * 1024
  static let maxHeaderValueWireBytes = 16 * 1024
  static let maxTotalHeaderWireBytes = 64 * 1024

  private static let tokenCharacters = Set(
    "!#$%&'*+-.^_`|~0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
  )

  static func normalize(_ headers: [String: String]) throws -> [String: String] {
    guard headers.count <= maxHeaderCount else {
      throw PolicyError("Native HTTP request has too many headers.")
    }

    var totalWireBytes = 0
    var normalized: [String: String] = [:]
    var normalizedNames = Set<String>()
    normalized.reserveCapacity(headers.count)
    normalizedNames.reserveCapacity(headers.count)
    for (rawName, rawValue) in headers {
      // Retain the bridge's historical behavior for an empty property name.
      if rawName.allSatisfy(\.isWhitespace) { continue }
      guard
        rawName.utf16.count <= maxHeaderNameCharacters,
        rawName.allSatisfy({ tokenCharacters.contains($0) })
      else {
        throw PolicyError("Native HTTP request has an invalid header name.")
      }
      let normalizedName = rawName.lowercased()
      guard normalizedNames.insert(normalizedName).inserted else {
        throw PolicyError("Native HTTP request has duplicate header names.")
      }
      guard rawValue.utf16.count <= maxHeaderValueCharacters else {
        throw PolicyError("Native HTTP request has an oversized header value.")
      }

      let value = try normalizeURLValue(normalizedName: normalizedName, value: rawValue)
      guard isPrintableASCII(value) else {
        throw PolicyError(
          "Native HTTP request header values must use printable ASCII."
        )
      }
      let valueWireBytes = value.utf8.count
      guard valueWireBytes <= maxHeaderValueWireBytes else {
        throw PolicyError("Native HTTP request has an oversized header value.")
      }
      // Header names and normalized values are ASCII at this point, so UTF-8
      // length is the exact HTTP/1 field-content byte length. Rechecking after
      // URL canonicalization prevents percent-encoding from expanding a small
      // Unicode input past the advertised per-value bound.
      totalWireBytes += rawName.utf8.count + valueWireBytes
      guard totalWireBytes <= maxTotalHeaderWireBytes else {
        throw PolicyError("Native HTTP request headers exceed the safety limit.")
      }
      // Sources never control Nemu's loopback-proxy credential. Validate and
      // count the field first so the reserved name cannot bypass bounds or
      // case-folded duplicate rejection.
      if normalizedName == "proxy-authorization" { continue }
      normalized[rawName] = value
    }
    return normalized
  }

  static func apply(
    _ headers: [String: String],
    to request: inout URLRequest
  ) throws {
    for (name, value) in try normalize(headers) {
      request.setValue(value, forHTTPHeaderField: name)
    }
  }

  private static func normalizeURLValue(
    normalizedName: String,
    value: String
  ) throws -> String {
    if isPrintableASCII(value) { return value }
    guard
      normalizedName == "referer" || normalizedName == "referrer" || normalizedName == "origin"
    else {
      return value
    }
    guard
      let url = URL(string: value),
      let scheme = url.scheme?.lowercased(),
      scheme == "http" || scheme == "https",
      url.host?.isEmpty == false,
      isPrintableASCII(url.absoluteString)
    else {
      throw PolicyError("Native HTTP request has an invalid URL header.")
    }
    return url.absoluteString
  }

  private static func isPrintableASCII(_ value: String) -> Bool {
    value.unicodeScalars.allSatisfy {
      $0.value == 0x09 || (0x20...0x7e).contains($0.value)
    }
  }

  private struct PolicyError: LocalizedError {
    let errorDescription: String?

    init(_ description: String) {
      errorDescription = description
    }
  }
}
